File: make_symlinks.rb
#!/usr/bin/env ruby

require 'optparse'
require 'pathname'
require 'pp'

force = false

ARGV.options do |opt|
  opt.banner = "Usage: #{File.basename($0)} [option]"
  opt.on("--help", "Print this message and quit.") { puts opt; exit 0 }
  opt.on("-f", "--force", "Force overwrite when file exist.") { force = true }
  opt.parse!
end

home = Pathname.new(ENV['HOME'])
dotfiles = Pathname.new(__FILE__).realpath.dirname.expand_path
puts "Source dir: #{dotfiles}"
dotfiles.each_entry do |entry|
  if (%w!. .. .svn .git! + [File.basename(__FILE__)]).include? entry.to_s
    puts "Skip: #{entry}"
    next
  end
  newlink = home+entry
  # リンク先が既に存在しているか？
  if newlink.exist?
    next if (newlink.symlink? and newlink.readlink.realpath == (dotfiles+entry).realpath)
    puts "#{newlink} is exists."
    next unless force
    # バックアップを作成
    bak = Pathname.new(newlink.to_s+'.bak')
    if bak.exist?
      if bak.directory?
        bak.rmtree
      else
        bak.unlink
      end
    end
    newlink.rename(bak)
  end
  newlink.make_symlink(dotfiles+entry)
  puts "Make symlink: #{newlink} -> #{dotfiles+entry}"
end

home.each_entry do |entry|
  file = (home+entry).expand_path
  if file.symlink? and not file.exist?
    file.unlink
    puts "Remove symlink: #{file}"
  end
end

File: make_symlinks.py
#!/usr/bin/env python

from os import path, listdir, remove, rename, symlink
from optparse import OptionParser
import re

parser = OptionParser()
parser.add_option('-f', '--force', action='store_true', dest='force',
                  help='Force overwrite when file exist.')
(options, args) = parser.parse_args()

def is_skip(entry):
    if entry in ['.', '..', '.svn', '.git', path.basename(__file__)]:
        return True
    if [1 for pattern in [r'#.*#', r'\.#.*', r'.*~'] if re.match(pattern, entry)]:
        return True
    return False

home = path.expanduser('~')
dotfiles = path.dirname(path.realpath(__file__))
print('Source dir: %s' % dotfiles)
for entry in listdir(dotfiles):
    if is_skip(entry):
        print('Skip: %s' % entry)
        continue
    srcfile = path.join(dotfiles, entry)
    newlink = path.join(home, entry)
    if path.exists(newlink):
        if (path.islink(newlink) and
            path.realpath(newlink) == path.realpath(srcfile)):
            continue
        print('%s is exists.' % newlink)
        if options.force:
            bak = newlink + '.bak'
            if path.exists(bak):
                remove(bak)
            rename(newlink, bak)
    else:
        print("Make symlink: %s -> %s" % (newlink, srcfile))
        symlink(srcfile, newlink)

for entry in listdir(home):
    homefile = path.join(home, entry)
    if path.islink(homefile) and not path.exists(homefile):
        remove(homefile)
        print('remove symlink: %s' % homefile)
